Vim tab names and separators

echuraev opened the pull request:
Implemented getting name of vim tabs. Added an option to chose how Vim tabs will be displayed. Added two options to set separator for active Vim tab.

echuraev commented:
@zefei Could you please review it and merge?

zefei commented on plugin/wintabs.vim:
```diff
@@ -69,6 +69,9 @@ call s:set('g:wintabs_ui_sep_inbetween', '|')
 call s:set('g:wintabs_ui_active_left', ' ')
 call s:set('g:wintabs_ui_active_right', ' ')
+call s:set('g:wintabs_ui_show_tab_name', 0)
+call s:set('g:wintabs_ui_active_lsepgroup', ' ')
```
let's call these `g:wintabs_ui_active_vimtab_[left|right]`

zefei commented on plugin/wintabs.vim:
```diff
@@ -69,6 +69,9 @@ call s:set('g:wintabs_ui_sep_inbetween', '|')
 call s:set('g:wintabs_ui_sep_rightmost', '|')
 call s:set('g:wintabs_ui_active_left', ' ')
 call s:set('g:wintabs_ui_active_right', ' ')
+call s:set('g:wintabs_ui_show_tab_name', 0)
```
let's call this `g:wintabs_ui_show_vimtab_name` to be consistent with the existing terms, though I don't think I named them well.

zefei commented on autoload/wintabs/ui.vim:
```diff
@@ -176,14 +203,18 @@ function! s:get_spaceline()
+    let name = ' '.name.' '
+
+    " get and normalize space name
+    let length += len(name)
```
this must happen between adding separator and adding highlight group name, as the highlight group shouldn't be counted towards character length

zefei commented on autoload/wintabs/ui.vim:
```diff
@@ -176,14 +203,18 @@ function! s:get_spaceline()
+      let name = g:wintabs_ui_active_lsepgroup.name.g:wintabs_ui_active_rsepgroup
       let name = '%#'.g:wintabs_ui_active_higroup.'#'.name.'%##'
     endif
+    let name = ' '.name.' '
```
this probably should be inside an `else`, because we don't want to add double spaces if `g:wintabs_ui_active_[l|r]sepgroup` isn't set. It might not be what you intended, but we should make sure that the old behavior is kept if `g:wintabs_ui_active_[l|r]sepgroup` isn't set.

echuraev commented:
I fixed all comments. And added one more value for `g:wintabs_ui_show_vimtab_name` and one more option `g:wintabs_ui_vimtab_separator`. You can see how it works on attached screenshots.

P.S. I don't know why but tab highlighting isn't working on my machine. With version from master it doesn't work too. So, I'm using `g:wintabs_ui_active_left` for highlighting current tab.

Default values:
![vim_win_defaults](https://cloud.githubusercontent.com/assets/5525113/21452486/45802a96-c918-11e6-9e80-f668fefe8746.png)

Set `g:wintabs_ui_active_vimtab_left` and `g:wintabs_ui_active_vimtab_right`:
![vim_win_active_vim_tabs](https://cloud.githubusercontent.com/assets/5525113/21452487/45819e58-c918-11e6-98f7-4f31e1ccb60a.png)

Set `g:wintabs_ui_show_vimtab_name = 1`:
![vim_win_show_wintab_name1](https://cloud.githubusercontent.com/assets/5525113/21452489/45833862-c918-11e6-91ac-073c7e78d47f.png)

Set `g:wintabs_ui_show_vimtab_name = 2`:
![vim_win_show_wintab_name2](https://cloud.githubusercontent.com/assets/5525113/21452490/45842cd6-c918-11e6-85a8-0655d8874ea7.png)

Set custom separator:
![vim_win_show_wintab_custom_sep](https://cloud.githubusercontent.com/assets/5525113/21452488/4581b460-c918-11e6-869b-ce3d7a2daa6f.png)

echuraev commented:
@zefei Could you please review my update?

Thank you in advance!

zefei commented on autoload/wintabs/ui.vim:
```diff
@@ -165,6 +165,37 @@ function! s:truncate_line(window, bufline, width)
+    let buflist = tabpagebuflist(a:n)
+    let winnr = tabpagewinnr(a:n)
+    let title = bufname(buflist[winnr - 1])
+    let title = split(title, "/")[-1]
```
This can error if you simply issue an `:tabnew` as the new vimtab has no buffers and `title` is empty.

zefei commented on autoload/wintabs/ui.vim:
```diff
@@ -176,13 +207,20 @@ function! s:get_spaceline()
-    let length += len(name)
+    " 2 is added to length of name because we will add two separators around
+    " name
+    let length += (len(name) + 2)
```
It's not guaranteed that user would use single character for separators, so let's put this logic `let length += len(name)` after each condition below (after line 220 and line 223).

zefei commented on doc/wintabs.txt:
```diff
@@ -200,6 +200,32 @@ default: ' '
+                This controls how Vim tabs will be displayed.
+                0: show a number of the tab.
+                1: show a name of the tab.
+                2: show number and name of the tab. (<number>:<tab_name>).
```
I'm against this option (number+name) as this intentionally isn't an option for the buffer list. The reason being that once the tab is named, its tab number becomes quite useless (looping through tabs or mouse click is faster than using tab number).

echuraev replied:
Yes, but it can be useful for someone who likes navigation as in tmux. For example, in tmux my tabs  is named like <tab_number>:<tab_name>. I enabled mouse mode in my tmux config but mostly I'm changing tabs using the number of the tab. It is more comfortable for me.

echuraev replied:
In additional, I can map command `tabn N` to some shortcut and next I'll be able to use number of tabs for faster switching between opened tabs.

echuraev replied:
@zefei Do you think that one additional option for customizing tab bar will be bad for whole plugin? If you think so then I'm going to remove this parameter in next commit.

zefei replied:
I think your use case is fair. I'm OK with this option, but please remove the format option `g:wintabs_ui_vimtab_separator` as I don't think it's good to implement it that way. A good way to make this feature would be to allow user to free form the tab name like `"%d: %n"`, but that's probably too much for this change.

echuraev replied:
Ok, thank you for the good feedback. I'll create ticket with this task. I think that it will be a good task to the future.

zefei commented on autoload/wintabs/ui.vim:
```diff
@@ -210,17 +214,14 @@ function! s:get_spaceline()
     " highlight current space
     if tab == tabpagenr()
       let name = g:wintabs_ui_active_vimtab_left.name.g:wintabs_ui_active_vimtab_right
+      let length += (len(name) + 2)
```
You no longer need `+2` here.

zefei commented on autoload/wintabs/ui.vim:
```diff
@@ -210,17 +214,14 @@ function! s:get_spaceline()
       let name = '%#'.g:wintabs_ui_active_higroup.'#'.name.'%##'
     else
       let name = ' '.name.' '
+      let length += (len(name) + 2)
```
Same here.

echuraev replied:
Sorry, I forgot to change it.

zefei commented:
Thanks for the PR!
